internal/controller/nonadminbackupstoragelocation_controller.go

```diff
@@ -377,6 +413,180 @@ func (r *NonAdminBackupStorageLocationReconciler) setFinalizerOnNaBSL(ctx contex
 	return false, nil
 }
 
+// ensureNonAdminRequest updates the NonAdminBackupStorageLocation object based on the
+// cluster admin's approval decision on the NonAdminBackupStorageLocationRequest object
+// and ensures Velero BackupStorageLocation and secret are deleted if the approval decision
+// is rejected
+func (r *NonAdminBackupStorageLocationReconciler) ensureNonAdminRequest(
+	ctx context.Context, logger logr.Logger, nabsl *nacv1alpha1.NonAdminBackupStorageLocation) (bool, error) {
+	veleroObjectsNACUUID := nabsl.Status.VeleroBackupStorageLocation.NACUUID
+
+	nabslRequest, err := function.GetNabslRequestByLabel(ctx, r.Client, r.OADPNamespace, veleroObjectsNACUUID)
+	if err != nil {
+		logger.Error(err, findSingleNABSLRequestError)
+		return false, err
+	}
+
+	var terminalErr error
+	var reason, message string
+
+	statusCondition := metav1.ConditionFalse
+	shouldRequeue := false
+	expectedPhase := nacv1alpha1.NonAdminPhaseNew
+	updatedRejectedCondition := false
+	updatedApprovedCondition := false
+
+	if !reflect.DeepEqual(nabslRequest.Status.VeleroBackupStorageLocationRequest.DeepCopy().RequestedSpec, nabsl.Spec.BackupStorageLocationSpec) {
+		message = "NaBSL Spec update not allowed. Changes will not be applied. Delete NaBSL and create new one with updated spec"
+		updatedRejectedCondition = meta.SetStatusCondition(&nabsl.Status.Conditions, metav1.Condition{
+			Type:    string(nacv1alpha1.NonAdminBSLConditionSpecUpdateApproved),
+			Status:  metav1.ConditionFalse,
+			Reason:  "BslSpecUpdateRejected",
+			Message: message,
+		})
+		statusCondition = metav1.ConditionTrue
+		expectedPhase = nabsl.Status.Phase
+		terminalErr = reconcile.TerminalError(errors.New(message))
+	} else if nabslRequest.Status.VeleroBackupStorageLocationRequest.NACUUID == constant.EmptyString || nabslRequest.Status.VeleroBackupStorageLocationRequest.NACUUID != nabsl.Status.VeleroBackupStorageLocation.NACUUID {
+		message = "NonAdminBackupStorageLocationRequest does not contain valid NAC UUID and can not be approved"
+		updatedRejectedCondition = meta.SetStatusCondition(&nabsl.Status.Conditions, metav1.Condition{
+			Type:    string(nacv1alpha1.NonAdminBSLConditionApproved),
+			Status:  metav1.ConditionFalse,
+			Reason:  "BslSpecUpdateRejected",
+			Message: message,
+		})
+		terminalErr = reconcile.TerminalError(errors.New(message))
+		expectedPhase = nacv1alpha1.NonAdminPhaseBackingOff
+	} else {
+		switch {
+		case nabslRequest == nil:
+			shouldRequeue = true
```

Comment:
I think you can remove this case, right ?
This condition will always be false (GetNabslRequestByLabel will return an err in nil case)

Reply:
Correct, actually `GetNabslRequestByLabel` is not returning error if nothing is found, works same in other Backup/Restore/NAB objects so don't want to change how the GetNabslRequestByLabel works, however this check is not required and added new one on the nabslRequest nil just afte GetNabslRequestByLabel so other parts of the function won't have nil object (which will in theory never happen, but better safe then sorry)